File: 5-Problem-A-Day/44. Wildcard Matching.java
class Solution {
    public boolean isMatch(String s, String p) {
        int sLen = s.length(), pLen = p.length();
        int sIdx = 0, pIdx = 0;
        int starIdx = -1, sTmpIdx = -1;
        while (sIdx < sLen) {
            if (pIdx < pLen && (p.charAt(pIdx) == '?' || p.charAt(pIdx) == s.charAt(sIdx))) {
                sIdx += 1;
                pIdx += 1;
            }
            else if (pIdx < pLen && p.charAt(pIdx) == '*') {
                starIdx = pIdx;
                sTmpIdx = sIdx;
                pIdx += 1;
            }
            else if (starIdx == -1)
                return false;
            else {
                pIdx = starIdx + 1;
                sIdx = sTmpIdx + 1;
                sTmpIdx = sIdx;
            }
        }
        for (int i = pIdx; i < pLen; i++)
            if (p.charAt(i) != '*')
                return false;
        return true;
    }

File: 2021/44. Wildcard Matching.cpp
class Solution {
public:
    bool isMatch(string s, string p) {
        int dp[s.length() + 1][p.length() + 1];
        memset(dp, 0, sizeof(dp));                
        dp[0][0] = 1;
        for (int i = 1; i <= s.length(); i++) {
            dp[i][0] = 0;
        }
        for (int j = 1; j <= p.length(); j++) {
            dp[0][j] = p[j - 1] == '*' && dp[0][j - 1] == 1;
        }
        
        for (int i = 1; i <= s.length(); i++) {
            for (int j = 1; j <= p.length(); j++) {
                if (p[j - 1] == '*') {
                    dp[i][j] = dp[i][j - 1] || dp[i - 1][j];
                } else {
                    dp[i][j] = (s[i- 1] == p[j - 1] || p[j - 1] == '?') && dp[i - 1][j - 1];
                }         
            }
        }
        
        return dp[s.length()][p.length()];
    }
};